source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is part of the Mooltipass Project (https://github.com/mooltipass).
+ * Copyright (c) 2019 Stephan Mathieu
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+/*!  \file     logic_totp.c
+*    \brief    General logic for TOTP algorithm based on RFC 6238 (https://tools.ietf.org/html/rfc6238)
+*    Created:  31/07/2020
+*    Author:   dnns01
+*/
+#include "logic_totp.h"
+
+/*! \fn     logic_totp_sha1_hmac(void *out, const void *key, const void *message)
+*   \brief  Calculate SHA1 HMAC Hash for given message using given key.
+*   \param  calculated hash, char[20] expected
+*   \param  key for SHA1 HMAC calculation, char[10] expected
+*   \param  message to be hashed, 8 bytes expected
+*   \return void
+*/
+static void logic_totp_sha1_hmac(void* out, const void *key, const void *message)
+{
+    br_hmac_key_context key_context;
+    br_hmac_context context;
+
+    br_hmac_key_init(&key_context, &br_sha1_vtable, key, LOGIC_TOTP_SHA1_HMAC_KEY_LEN);
+    br_hmac_init(&context, &key_context, 0);
+    br_hmac_update(&context, message, LOGIC_TOTP_SHA1_HMAC_MESSAGE_LEN);
+    br_hmac_out(&context, out);
+}
+
+
+/*! \fn     logic_totp_get_time()
+*   \brief  Retrieves time from RTC and converts calendar back to unix timestamp
+*   \return current unix timestamp
+*/
+static time_t logic_totp_get_time(void)
+{
+    calendar_t calendar;
+    struct tm datetime;
+
+    timer_get_calendar(&calendar);
+
+    datetime.tm_sec = calendar.bit.SECOND;
+    datetime.tm_min = calendar.bit.MINUTE;
+    datetime.tm_hour = calendar.bit.HOUR;
+    datetime.tm_mday = calendar.bit.DAY;
+    datetime.tm_mon = calendar.bit.MONTH - 1;
+    datetime.tm_year = calendar.bit.YEAR + 2000 - 1900;
+
+    // Assumption: If running in the emulator, gmtime is used. Therefore we call _mkgmtime to get the current timestamp, whereas
+    // the real device is running on UTC, therefore mktime is used to get the timestamp.
+#ifndef EMULATOR_BUILD
+    return mktime(&datetime);
+#else
+    return _mkgmtime(&datetime);
+#endif
+}
+
+
+/*! \fn     logic_totp_base32_decode_key(char* decoded_key, const char* encoded_key)
+*   \brief  Decodes the given Base32 encoded key
+*   \param  decoded key, char[10] expected
+*   \param  Base32 encoded key, char[16] expected
+*   \return void
+*/
+static void logic_totp_base32_decode_key(char* decoded_key, const char* encoded_key)
```

Comment:
if you wish to, for the decoded key you may use uint8_t :)

Reply:
I would suggest that we pass in the binary key from outside (Moolticute app) so we don't have to deal with base32 encoding and decoding on the minible....

Reply:
that's a very good point